challenge_demo/test/grammar_test.py

```diff
@@ -0,0 +1,130 @@
+        cls.task_manager = TaskManager(robot=Mockbot())
+
+        # ToDo: Should we make everything recursive???
+
+        # Rule corresponding to the grammar target
+        root_rule = cls.parser.rules[cls.knowledge.grammar_target]
+
+        # In principle, a list of actions can be provided to the actions server
+        # Therefore, we need to take to steps to list all actions
+        # I.e., the following actions_list_target refers to the grammar target
+        # containing a list of actions
+        actions_list_target = ""
+        for option in root_rule.options:
+            if "actions" not in option.lsemantic:
+                continue
+            actions_list_target = option.conjuncts[0].name  # This is too hardcoded
+            break
+
+        assert actions_list_target, "Cannot derive actions from grammar"
+        actions_list_rule = cls.parser.rules[actions_list_target]
+        assert len(actions_list_rule.options) == 1, \
+            "Don't know how to continue if the actions list rule has multiple options"
+        assert len(actions_list_rule.options[0].conjuncts) == 1, \
+            "Don't know how to continue if the actions options list rule has multiple conjunctions"
+
+        # Now, we need to get all separate options
+        actions_target = actions_list_rule.options[0].conjuncts[0].name
+
+        cls.actions_rule = cls.parser.rules[actions_target]
+
+    def test_grammar(self):
+
+        test_results = {}
+        for option in self.actions_rule.options:
+
+            result_str = ""
+
+            # For each conjunct, we want to have a viable option
+            for conjunct in option.conjuncts:
+                result_str += self._resolve_conjunct(conjunct.name)
+            result_str = result_str.rstrip(" ")
+
+            # Now, we can parse the result string to get the action description
+            actions_definition = self.parser.parse_raw(
+                target=self.knowledge.grammar_target,
+                words=result_str,
+            )
+            self.assertIn(
+                "actions",
+                actions_definition,
+                "'actions' not in actions definition, don't know what to do"
+            )
+
+            # Test the action description with the task manager
+            try:
+                test_result = self.task_manager.set_up_state_machine(
+                    recipe=actions_definition["actions"],
+                )  # type: ConfigurationResult
+            except Exception as e:
+                test_result = ConfigurationResult()
+                test_result.message = "Configuration crashed: {}".format(e.message)
+            test_results[result_str] = TMTestResult(actions_definition["actions"], test_result)
+
+        failed_test_results = {}
+        for action, test_result in test_results.iteritems():
+            if test_result.config_result.succeeded:
+                print("Configuration of '{}' succeeded".format(action))
```

Comment:
Why prints?

Reply:
What else? If I use rospy I need a rostest instead of a pytest. This is extremely inconvenient. I don't want print statements in my library code but in tests I do find it acceptable.